### DISCUSSION TITLE
Add a string to nlohmann::json without copying?

### DISCUSSION THREAD

**Discussion by bhaller** [nlohmann/json#3105] [Q&A] [answered]
Hi!  This question may be related to #2517, but is much simpler I think/hope.

I've got a C string in a buffer, NUL-terminated.  It is very large, perhaps as large as 1e10 characters.  I would therefore like to get it into JSON without copying.  I am happy to donate the (malloced) buffer to the JSON object.  Is there a reasonably straightforward way to do so?  At present I have:

`metadata.emplace("reference_sequence", std::string(buffer));`

where `metadata` is a new `nlohmann::json` object.  My understanding is that the use of `emplace()` should avoid an extra copy being made, as I guess would occur if I did:

`metadata["reference_sequence"] = std::string(buffer);`

But if I understand correctly, the `std::string()` constructor still copies the contents of `buffer`, by definition.  Is there a way to avoid this?  I.e., does `nlohmann::json` have any way to say "emplace characters from this malloced buffer, of this length, as a string value, taking ownership of the buffer", rather than needing a `std::string`?

EDIT: In writing this question, it occurred to me that maybe I could make a `std::string`, `resize()` that string to the needed size, and throw my data into that string's `data()` buffer, and then `emplace()` the result into `nlohmann::json`.  That does not seem to be possible, however, because `std::string`'s `data()` method returns a `const char *`.  But the discussion at https://stackoverflow.com/a/14291203/2752221 indicates that I *can* get a non-const pointer from `std::string` to do what I want, using `operator[]`, and that is considered legal according to the standard (go figure) as long as I don't touch the terminating NUL, which is fine.  So if that works as hoped, then maybe this question is academic; I'm still curious to hear the answer, though, if somebody feels like answering it.  Thanks!

**Comment by nlohmann** [comment, marked as answer]
No, unfortunately, this does not work. :-/

The library uses `std::string` under the hood and it seems to be impossible to BYO buffer. You *could* solve this issue by templating the library with your own string type, but we have little experience with this.

**Reply by bhaller** [reply]
I see, thanks.  I guess the way I found is best, then – use a `std::string` as the buffer on my end, and then `emplace()` it into the json object.  I implemented that and it seems to work fine, although I haven't tested whether it actually removes all copying of the buffer data in practice.

**Reply by gregmarr** [reply]
@bhaller You need to wrap your string object in `std::move()` when calling `emplace()` so it is passed as `std::string &&` instead of `std::string &` to have the internal `std::string()` take over the buffer.

**Reply by bhaller** [reply]
@gregmarr Oh!  I thought using `emplace()` implied move semantics automatically; I didn't realize a `std::move()` was needed as well!  Yikes; I will need to check this across my whole code base, I guess.  I'm still a rather beginner C++ programmer, as you can see.  Thanks for the tip!

**Reply by gregmarr** [reply]
`emplace()` means construct the destination object in place by calling a constructor using these parameters rather than using the default constructor and copying or moving in the fully constructed object.  It says nothing about the value category of the parameters themselves when passed to the constructor.

**Reply by gregmarr** [reply]
j.push_back(json(my_string));

copies `my_string` into a temporary `json`, then default constructs a new `json` at the end of the vector and moves the temporary into it.

    j.push_back(json(std::move(my_string)));

moves `my_string` into a temporary `json`, then default constructs a new `json` at the end of the vector and moves the temporary into it.

    j.emplace(my_string);

copies `my_string` directly into a new `json` at the end of the vector.

    j.emplace(std::move(my_string));

moves `my_string` directly into a new `json` at the end of the vector.

**Reply by bhaller** [reply]
Thank you for the clarification!
